kong/tools/wrpc.lua

```diff
@@ -0,0 +1,685 @@
+      local keyword, identifier = nextline:match("^%s*(%a+)%s+(%w+)")
+      if keyword and identifier then
+
+        if keyword == "service" then
+          scope_name = identifier;
+
+        elseif keyword == "rpc" then
+          identifier = scope_name .. "." .. identifier
+        end
+
+        local type_annotations = annotations[keyword]
+        if type_annotations then
+          local tag_key, tag_value = annotation:match("^%s*(%S-)=(%S+)%s*$")
+          if tag_key and tag_value then
+            tag_value = tag_value
+            local tags = type_annotations[identifier] or {}
+            type_annotations[identifier] = tags
+            tags[tag_key] = tag_value
+          end
+        end
+      end
+    end
+  end
+  proto_f:close()
+
+  grpc.each_method(service_fname, function(_, srvc, mthd)
+    assert(srvc.name)
+    assert(mthd.name)
+    local rpc_name = srvc.name .. "." .. mthd.name
+
+    local service_id = assert(annotations.service[srvc.name] and annotations.service[srvc.name]["service-id"])
+    local rpc_id = assert(annotations.rpc[rpc_name] and annotations.rpc[rpc_name]["rpc-id"])
+    local rpc = {
+      name = rpc_name,
+      service_id = tonumber(service_id),
+      rpc_id = tonumber(rpc_id),
+      input_type = mthd.input_type,
+      output_type = mthd.output_type,
+    }
+    self.methods[service_id .. ":" .. rpc_id] = rpc
+    self.methods[rpc_name] = rpc
+  end, true)
+end
+
+--- returns the method defintion given either:
+--- pair of IDs (service, rpc) or
+--- rpc name as "<service_name>.<rpc_name>"
+function wrpc_service:get_method(srvc_id, rpc_id)
+  local rpc_name
+  if type(srvc_id) == "string" and rpc_id == nil then
+    rpc_name = srvc_id
+  else
+    rpc_name = tostring(srvc_id) .. ":" .. tostring(rpc_id)
+  end
+
+  return self.methods[rpc_name]
+end
+
+--- sets a service handler for the givern rpc method
+--- @param rpc_name string Full name of the rpc method
+--- @param handler function Function called to handle the rpc method.
+--- @param response_handler function Fallback function called to handle responses.
+function wrpc_service:set_handler(rpc_name, handler, response_handler)
+  local rpc = self:get_method(rpc_name)
+  if not rpc then
+    return nil, string.format("unknown method %q", rpc_name)
+  end
+
+  rpc.handler = handler
+  rpc.response_handler = response_handler
+  return rpc
+end
+
+
+--- Part of wrpc_peer:call()
+--- If calling the same method with the same args several times,
+--- (to the same or different peers), this method returns the
+--- invariant part, so it can be cached to reduce encoding overhead
+function wrpc_service:encode_args(name, ...)
+  local rpc = self:get_method(name)
+  if not rpc then
+    return nil, string.format("unknown method %q", name)
+  end
+
+  local num_args = select('#', ...)
+  local payloads = table.new(num_args, 0)
+  for i = 1, num_args do
+    payloads[i] = assert(pb.encode(rpc.input_type, select(i, ...)))
```

Comment:
Encode every arguments with the same `input_type` seems a bug.

Reply:
the RPC definitions in .proto files only specify one input type.   still, the frame structure supports repeated payloads.  a single call could specify several items of the same time instead of using an array.

Reply:
Maybe I misinterpreted the comment in the wrpc.proto:
```protobuf
  // payloads is an array representing the request or response data of an RPC.
  // A request message MAY contain multiple elements where each element represents
  // an argument to the RPC call. The order of the elements MUST correspond to
  // the order of arguments in the function call.
  // A response message MUST contain a single payload.
  // Use a wrapper type for RPCs which contain multiple responses.
  // Unless otherwise specified by the Service or RPC, the encoding method in
  // use is PROTO3.
  //
  // Note: This results in double proto3 encoding. Once encoding of payloads,
  // and then encoding of the entire message. We acknowledge that there are
  // some overheads here and they will be addressed in a future iteration.
  repeated bytes payloads = 9;
```

By "each element represents an argument" I thought it was for every member of the request message (like "hash" and "plugins" in the previous comment).
Also, if an RPC call receives repeated messages, it cannot use repeat like this:

```proto
  rpc SomeCall(repeated SomeCallRequest) returns (SomeCallResponse);
```

But to use something like:

```proto
Message RepeatedSomeCallRequest {
    repeated SomeCallRequest arguments = 1;
}
```

So I don't think it makes sense to encode an array of payloads...

Reply:
> Also, if an RPC call receives repeated messages, it cannot use repeat like this:
>
>  rpc SomeCall(repeated SomeCallRequest) returns (SomeCallResponse);

right, the syntax doesn't support this.  but in the current implementation, the "repeated" is implied, so this could be done:

```protobuf
rpc SomeCall(SomeCallRequest) returns (someCallResponse);

message SomeCallRequest {
    string name = 1;
    string value = 2;
}
```
and in Lua:
```lua
   peer:call("serv.SomeCall", {name="one", value=1}, {name="two", value=2}, {name="three", value=3})
```

and it would result in a single call with three payload elements of the same type.

But I agree that it's not an important usecase, and the structure you describe would be just as good.

Reply:
@hbagdi May you clarify this? Do we really want to implement it this way?